taichi/backends/opengl/codegen_opengl.cpp

```diff
@@ -486,13 +486,14 @@ class KernelGen : public IRVisitor {
     const auto &element_shape = stmt->element_shape;
     std::vector<std::string> size_var_names;
     std::vector<std::string> element_shape_size_var_names;
-    enum ExternalArrayLayout { layout_AOS = 0, layout_SOA = 1 };
-    const auto layout = stmt->element_dim <= 0 ? layout_AOS : layout_SOA;
+
+    const auto layout = stmt->element_dim <= 0 ? ExternalArrayLayout::kAOS
+                                               : ExternalArrayLayout::kSOA;
 
     if (element_shape.size() > 0) {
       int elem_beg = 0;
       int elem_end = 0;
-      if (layout == layout_SOA) {
+      if (layout == ExternalArrayLayout::kSOA) {
```

Comment:
IIUC, once element_shape is no longer passed into extra_args, we can  remove this part one...? (Metal now has the same stuff due to the Ndarray support) cc @turbo0628

Reply:
Yup we briefly discussed this in the afternoon and @turbo0628 will get rid of these hacks later (still need some more cleanup before that